theano/tensor/elemwise.py

```diff
@@ -1059,35 +1161,60 @@ def _c_all(self, node, nodename, inames, onames, sub):
             alloc += cgen.make_checks([list(range(nnested))], [odtype],
                                       dict(sub, lv0=oname))
-        olv_index = i  # index of the last output
-
-        # We loop over the "aliased" outputs, i.e., those that are
-        # inplace (overwrite the contents of one of the inputs) and
-        # make the output pointers point to theur corresponding input
-        # pointers.
-        for output, oname in izip(aliased_outputs, aliased_onames):
-            olv_index = inputs.index(dmap[output][0])
-            iname = inames[olv_index]
-            # We make the output point to the corresponding input and
-            # decrease the reference of whatever the output contained
-            # prior to this
-            alloc += """
-            if (%(oname)s) {
-                Py_XDECREF(%(oname)s);
-            }
-            %(oname)s = %(iname)s;
-            Py_XINCREF(%(oname)s);
-            """ % locals()
-            # We alias the scalar variables
-            defines += "#define %(oname)s_i %(iname)s_i" % locals()
-            undefs += "#undef %(oname)s_i" % locals()
 
-        # Note: here, olv_index is either the index of the last output
-        # which is allocated, OR, if there are any aliased outputs,
-        # the index of the last of these aliased outputs.
+        # index of the last output
+        olv_index = i
+        task_code = ""
+
+        if 'params' in sub:
+
+            # All variables names, inputs followed by outputs
+            vnames = _inames+_onames
+
+            # We declare another array to easily access the scalar variables
+            # from their respectives index
+            defines = """
+            void *refVar[%i];
+            """ % len(vnames)
+            for (i, name) in enumerate(vnames):
+                defines += """
+                refVar[%(i)i] = &%(name)s_i;
+                """ % locals()
+
+            # We loop over the "aliased" outputs, i.e., those that are
+            # inplace (overwrite the contents of one of the inputs) and
+            # make the output pointers point to their corresponding input
+            # pointers.
+
+            for output, oname, i in izip(node.outputs, onames,
+                                         range(len(onames))):
+                idtype = output.type.dtype_specs()[1]
+
+                # Temporary hack waiting to find a better solution
+                # We would like the C code to execute something similar
+                if(i in self.inplace_pattern):
+                    name = _inames[self.inplace_pattern[i]]
+                else:
+                    name = _onames[i]
+                defines += """
+                    typedef typeof(%(name)s_i) t%(i)i;
+                """ % locals()
+                # We alias the scalar variables
+                defines += """
+                #define %(oname)s_i *(t%(i)i*) (refVar[outIndexes[%(i)i]])
```

Comment:
Basically, we would be able to retrieve the type of the variable where the refVar array points to and typedef it to t%(i) so that the casting is properly done (else it will lead to segfaults). However I was not able to find a solution relying only on C code. I think I'm thinking too much object oriented and not enough in C.

That or find an other implementation than macro to alias the inplace variables.

In order to make the test pass I've committed a temporary hack that will be removed from history once a solution is found.